src/lib/Libifl/auth.c

```diff
@@ -369,31 +369,41 @@ is_valid_encrypt_method(char *method)
  *	tcp_send_auth_req - encodes and sends PBS_BATCH_Authenticate request
  *
  * @param[in] sock - socket descriptor
+ * @param[in] port - parent port in pbs_iff (only used in resvport auth) else 0
+ * @param[in] user - authenticating user name
+ * @param[in] auth_method - auth method name
+ * @param[in] encrypt_method - encrypt method name
  *
  * @return	int
  * @retval	0 on success
  * @retval	-1 on error
  */
 int
-tcp_send_auth_req(int sock, unsigned int port, char *user)
+tcp_send_auth_req(int sock, unsigned int port, char *user, char *auth_method, char *encrypt_method)
 {
 	struct batch_reply *reply;
 	int rc;
-	int am_len = strlen(pbs_conf.auth_method);
-	int em_len = strlen(pbs_conf.encrypt_method);
+	int am_len;
+	int em_len = encrypt_method ? strlen(encrypt_method) : 0;
 
+	if (auth_method == NULL || *auth_method == '\0') {
+		/* auth method can't be null or empty string */
+		pbs_errno = PBSE_INTERNAL;
```

Comment:
Just a question. Earlier this function used to return only PBSE_SYSTEM in case of any error. Now we are returning PBSE_INTERNAL if PBS_CONF_ENCRYPT_METHOD is not provided as an environment variable. It looks like the caller of this function does not check pbs_errno in case of a failure and say the exact reason why it has failed and always give the error "Failed to send auth request". Do we want to take care of this at the caller ?

Reply:
No, it will be taken care in pbs_connect() (aka caller of caller to this func), pbs_connect() is already printing message: `"auth: error returned: %d\n", pbs_errno`

Reply:
Thanks @hirenvadalia. Good that we are checking pbs_errno in pbs_connect(). Along with the error you mentioned above we also get(if errbuff not null) "auth:Failed to send auth request" which I think it is very much general, may be if we could see the second error as "auth:<env_ name> is not provided." it would be easier for the caller of caller to know that authentication is failing because the corresponding environement variable is not provided.
What do you think ?

Reply:
I don't think we need that, as pbs_connect() is just printing error code and message return by engage_client_auth(), pbs_connect() doesn't care what auth methods it is using underneath.
Also this setting env is only comes into picture when auth is resvport, so its not generic.

Reply:
I am leaving it up to you @hirenvadalia. My only concern was that if we do the above,  one can easily know that authentication is failing due to the lack of the corresponding environment variable (even if resvport auth is less freqently used) instead of providing generic error.

Reply:
@suresh-thelkar he is also printing the setenvironmentvar() api failure into stderr, so that would be seen as well.

Reply:
I only saw windows counter part for the above i.e. SetEnvironmentVariable() where we are doing this and didn't find a LINUX counterpart. @subhasisb/ @hirenvadalia, Am I missing anything here ?

Reply:
@suresh-thelkar for *nix we don't need to set any env explicitly in current process, as `popen()` can understand `env=var <pbs_iff cmd>` syntax and that's what we use in *nix, please see `#else` part of `#ifdef WIN32`

Reply:
Got it @hirenvadalia. Thanks.